File: aimvision-mobile/plugins/phone-frame-sink/android/AVPhoneFrameSinkBridge.kt
/*
 * AVPhoneFrameSinkBridge.kt
 *
 * Slice 3c of ADR-0009 — Kotlin/JNI bridge into the Rust
 * `aimvision-camera-phone` crate's C ABI (see
 * ../../../../aimvision-camera-core/crates/aimvision-camera-phone/include/
 * aimvision_camera_phone.h).
 *
 * # Why this is a singleton with a `loadLibrary` try/catch
 *
 * The Rust crate ships as `libaimvision_camera_phone.so`. JNI requires
 * `Java_<package>_<class>_<method>`-named exports, which the bare C ABI
 * does NOT have — so a small JNI C-shim (`libaimvision_camera_phone_jni.so`)
 * is required between Kotlin and the C ABI. That shim is a follow-up
 * sub-slice (3c-android-jni); until it lands, `System.loadLibrary`
 * raises `UnsatisfiedLinkError` and the bridge reports unavailable —
 * the Vision Camera plugin falls back to the metadata-only path from
 * slice 3b, matching the iOS dlsym fallback in `AVPhoneFrameSinkBridge.swift`.
 *
 * # External function naming
 *
 * The `external fun` signatures here are the JNI-side contract the
 * follow-up shim must implement. They mirror the C ABI 1:1 in argument
 * order; the shim translates the `Long` opaque handle into a
 * `AimvisionPhoneCamera*` and forwards the call.
 *
 * # Lifecycle
 *
 * Process-singleton: one `PhoneCamera` per app, never freed (OS cleans up
 * on process exit). Matches the iOS bridge.
 */

package com.aimvision.app.phoneframesink

object AVPhoneFrameSinkBridge {
    @Volatile
    private var loaded: Boolean = false

    @Volatile
    private var cameraHandle: Long = 0L

    @Volatile
    private var monotonicSeq: Long = 0L

    init {
        try {
            // The JNI shim shipped from `aimvision-camera-phone-jni`
            // (follow-up sub-slice) is what hosts the `Java_..._native*`
            // exports declared below. Until that ships this loadLibrary
            // raises and we stay in the "unavailable" state.
            System.loadLibrary("aimvision_camera_phone_jni")
            cameraHandle = nativeNew("phone-0", 64L, 32L)
            loaded = cameraHandle != 0L
        } catch (_: UnsatisfiedLinkError) {
            // Expected on day zero — bridge reports unavailable, plugin
            // falls back to metadata-only output.
            loaded = false
            cameraHandle = 0L
        } catch (_: Throwable) {
            // Defensive: any other init failure also fails-soft. A
            // broken bridge must not crash the app.
            loaded = false
            cameraHandle = 0L
        }
    }

    /**
     * True iff the JNI shim was loaded AND the Rust `_new` call returned
     * a non-zero handle. When false the plugin must not call
     * [pushFrameMetadata].
     */
    @JvmStatic
    fun isAvailable(): Boolean = loaded

    /**
     * Per-camera wrap-safe monotonic frame counter. Single worklet thread
     * caller; the `@Synchronized` is defensive against future migrations
     * to a multi-thread emitter.
     */
    @JvmStatic
    @Synchronized
    fun nextMonotonicSeq(): Long {
        monotonicSeq += 1L
        return monotonicSeq
    }

    /**
     * Push one frame metadata record to Rust. Returns `true` on a
     * successful push, `false` when the bridge is unavailable.
     */
    @JvmStatic
    fun pushFrameMetadata(
        formatRaw: Int,
        handleId: Long,
        width: Int,
        height: Int,
        timestampNs: Long,
        monotonicSeq: Long,
    ): Boolean {
        if (!loaded || cameraHandle == 0L) return false
        nativePushFrame(cameraHandle, formatRaw, handleId, width, height, timestampNs, monotonicSeq)
        return true
    }

    /** Cumulative dropped-frames counter from the Rust ring buffer. */
    @JvmStatic
    fun droppedFramesCount(): Long {
        if (!loaded || cameraHandle == 0L) return 0L
        return nativeDroppedFrames(cameraHandle)
    }

    // JNI exports — implemented by the follow-up JNI shim
    // `libaimvision_camera_phone_jni.so`. The shim translates the opaque
    // `Long` handle into a `AimvisionPhoneCamera*` and forwards to the
    // C ABI in `aimvision_camera_phone.h`.
    @JvmStatic
    private external fun nativeNew(id: String, frameCapacity: Long, audioCapacity: Long): Long

    @JvmStatic
    @Suppress("UnusedPrivateMember")
    private external fun nativeFree(handle: Long)

    @JvmStatic
    private external fun nativePushFrame(
        handle: Long,
        formatRaw: Int,
        handleId: Long,
        width: Int,
        height: Int,
        timestampNs: Long,
        monotonicSeq: Long,
    ): Boolean

    @JvmStatic
    private external fun nativeDroppedFrames(handle: Long): Long
}

File: aimvision-mobile/plugins/phone-frame-sink/ios/AVPhoneFrameSinkBridge.swift
//
//  AVPhoneFrameSinkBridge.swift
//  Slice 3c of ADR-0009 — Swift side of the FFI bridge into the Rust
//  `aimvision-camera-phone` crate's C ABI (see
//  ../../../../aimvision-camera-core/crates/aimvision-camera-phone/include/
//  aimvision_camera_phone.h).
//
//  Symbol resolution strategy: dlsym against the main bundle. When the
//  dev has run `cargo build` and bundled the static library into the
//  Xcode target, the C ABI symbols are exported into the main binary,
//  dlsym finds them, and frame pushes start going to Rust. Until then
//  `isAvailable` reports `false` and the plugin's `callback` falls back
//  to the metadata-only path from slice 3b — this keeps the app running
//  without the Rust library on day-zero of integration.
//
//  Why dlsym (not direct linking from Swift):
//  - We want a single source file that compiles whether or not the Rust
//    library is present in the dev's build. Direct extern declarations
//    would fail to link without the library; dlsym defers that to a
//    runtime check we can fall back from.
//  - The performance penalty is paid once at init, not per-frame.

import Foundation

private typealias FFI_New = @convention(c) (
    UnsafePointer<CChar>?, Int, Int
) -> UnsafeMutableRawPointer?
private typealias FFI_Free = @convention(c) (UnsafeMutableRawPointer?) -> Void
private typealias FFI_PushFrame = @convention(c) (
    UnsafeMutableRawPointer?, UInt32, UInt64, UInt32, UInt32, UInt64, UInt64
) -> Bool
private typealias FFI_DroppedFrames = @convention(c) (UnsafeMutableRawPointer?) -> UInt64

/// Singleton FFI bridge to the Rust `PhoneCamera`. The lifecycle is
/// process-lifetime: we never call `_free` because the camera is meant
/// to live as long as the app does. The OS reclaims everything on exit.
@objc(AVPhoneFrameSinkBridge)
public final class AVPhoneFrameSinkBridge: NSObject {
    @objc public static let shared = AVPhoneFrameSinkBridge()

    /// `true` iff the C ABI symbols were resolved at process start AND
    /// the Rust `_new` call returned a non-NULL handle. When false the
    /// plugin must not call `pushFrame`.
    @objc public let isAvailable: Bool

    private let cameraHandle: UnsafeMutableRawPointer?
    private let pushFrame: FFI_PushFrame?
    private let droppedFrames: FFI_DroppedFrames?

    private override init() {
        // dlopen(NULL, ...) returns a handle to the running process — all
        // symbols statically linked into the main binary are visible
        // through it. RTLD_LAZY is fine here; we resolve in init.
        guard let proc = dlopen(nil, RTLD_LAZY) else {
            self.cameraHandle = nil
            self.pushFrame = nil
            self.droppedFrames = nil
            self.isAvailable = false
            super.init()
            return
        }
        defer { dlclose(proc) }

        guard let newSym = dlsym(proc, "aimvision_phone_camera_new"),
              let pushSym = dlsym(proc, "aimvision_phone_camera_push_frame"),
              let droppedSym = dlsym(proc, "aimvision_phone_camera_dropped_frames") else {
            self.cameraHandle = nil
            self.pushFrame = nil
            self.droppedFrames = nil
            self.isAvailable = false
            super.init()
            return
        }

        let newFn = unsafeBitCast(newSym, to: FFI_New.self)
        let pushFn = unsafeBitCast(pushSym, to: FFI_PushFrame.self)
        let droppedFn = unsafeBitCast(droppedSym, to: FFI_DroppedFrames.self)

        // Per-process singleton: capacity 64 frames ≈ 2 s @ 30 fps; 32
        // audio chunks ≈ 320 ms @ 10 ms chunks. Matches the slice-3a
        // tests' default scaling.
        let handle: UnsafeMutableRawPointer? = "phone-0".withCString { ptr in
            newFn(ptr, 64, 32)
        }

        self.cameraHandle = handle
        self.pushFrame = pushFn
        self.droppedFrames = droppedFn
        self.isAvailable = handle != nil
        super.init()
    }

    /// Push one frame to Rust. Returns `true` on a successful push,
    /// `false` if the bridge is unavailable. The bool the C ABI returns
    /// (whether a frame was evicted to make room) is exposed through the
    /// `droppedFramesCount` accessor — eviction is normal under
    /// backpressure and doesn't fail the push.
    @objc public func pushFrameMetadata(
        formatRaw: UInt32,
        handleId: UInt64,
        width: UInt32,
        height: UInt32,
        timestampNs: UInt64,
        monotonicSeq: UInt64
    ) -> Bool {
        guard isAvailable,
              let cameraHandle = cameraHandle,
              let pushFrame = pushFrame else {
            return false
        }
        _ = pushFrame(cameraHandle, formatRaw, handleId, width, height, timestampNs, monotonicSeq)
        return true
    }

    /// Read the Rust-side cumulative dropped-frames counter. Returns 0
    /// when the bridge is unavailable.
    @objc public var droppedFramesCount: UInt64 {
        guard isAvailable,
              let cameraHandle = cameraHandle,
              let droppedFrames = droppedFrames else {
            return 0
        }
        return droppedFrames(cameraHandle)
    }
}
